src/Elasticsearch/Endpoints/AbstractEndpoint.php

```diff
@@ -60,14 +64,18 @@
     /**
      * @param Transport $transport
      */
-    public function __construct($transport)
+    public function __construct($transport, RequestFactory $requestFactory = null)
     {
         $this->transport = $transport;
+        $this->requestFactory = $requestFactory;
     }
 
     /**
      * @throws \Exception
      * @return array
+     *
+     * @deprecated Performing a request directly on an endpoint is deprecated since 2.2 and will be removed in 3.0,
+     *             you should instead use the createRequest method
      */
     public function performRequest()
```

Comment:
PSR7 is only about the request, response, url, stream and file used in an Http context, it tell nothings about a client and how it must be send.

> So in light of that, perhaps it makes more sense to just write shims that can convert to/from PSR-7, instead of rewiring all the guts?

That is my intention for this PR, do not go beyond the PSR7, so it will be easy to accept this withtout breaking too much stuff.

About the client options thing, i may have jump too far by saying this is not possible anymore.
In fact, it may need extra works to decouple parameters from the `$params` and pass them to the transport, but should not be a big problem in this PR. I was just pointing the fact that the endpoint does not have the information about the client anymore, so it will induce more work for end user when switching (as he will need to separate the parameters of the endpoint from the client, but IMO this is a good thing as they are unrelated).

About guzzlering my intention is to propose, after this one, another pull request to use [httplug](http://docs.php-http.org/en/latest/). 

Like you said : 

> don't really see the need to support umpteen different (nearly identical) HTTP libs. 

Your are right, That's is cleary not a job for an api client such as elasticsearch-php, and that's what this library is about, providing a simple contract for http clients and providing adapters for all existing implementations (we don't have the guzzlering implementation, but AFAIK it shouldn't be so difficult to add).

The benefit of supporting multiple client through the same interface is to let the final user choose the implementating but still having the same contract. 
Then it allows using the same contract for differents api (there is others libraries who are in the process to switch to httplug) and this is clearly a win as the final user can have all the wanted behavior (such as logging, profiling, caching, etc ...) on this contract and they can be used everywhere.

There is other things in httplug that will help to reduce your maintenance cost, but don't want to talk too much of this here as it's not the subject for this PR (will do a long explanation of pros / cons in the following PR :) )